Wikipedia/Code/ArticleContainerViewController.swift

```diff
@@ -18,26 +18,154 @@ class ArticleContainerViewController: ViewController {
+    
+    private lazy var toolbarController: ArticleToolbarController = {
+        return ArticleToolbarController(toolbar: toolbar, delegate: self)
+    }()
+
     private let schemeHandler: SchemeHandler
+    private let dataStore: MWKDataStore
+    private let authManager: WMFAuthenticationManager = WMFAuthenticationManager.sharedInstance // TODO: DI?
+    private let alertManager: WMFAlertManager = WMFAlertManager.sharedInstance
     private let cacheController: CacheController
     private let articleURL: URL
+    private let article: WMFArticle
     private let language: String
     
+    private var leadImageHeight: CGFloat = 210
+    
+    @objc convenience init?(articleURL: URL, dataStore: MWKDataStore, theme: Theme) {
+        
+        guard let cacheController = dataStore.articleCacheControllerWrapper.cacheController else {
+            return nil
+        }
+        
+        self.init(articleURL: articleURL, dataStore: dataStore, cacheController: cacheController, theme: theme)
+    }
+    
+    init?(articleURL: URL, dataStore: MWKDataStore, schemeHandler: SchemeHandler = SchemeHandler.shared, cacheController: CacheController, theme: Theme) {
+        guard
+            let language = articleURL.wmf_language,
+            let article = dataStore.fetchOrCreateArticle(with: articleURL)
+        else {
+            return nil
+        }
+        
+        self.articleURL = articleURL
+        self.dataStore = dataStore
+        self.article = article
+        self.language = language
+        self.schemeHandler = schemeHandler
+        self.schemeHandler.articleCacheController = cacheController
+        self.cacheController = cacheController
+        
+        super.init(theme: theme)
+    }
+    
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+    }
+    
+    // MARK: WebView
+    
+    static let webProcessPool = WKProcessPool()
+    
+    lazy var messagingController: ArticleWebMessagingController = ArticleWebMessagingController(delegate: self)
+    
+    lazy var webViewConfiguration: WKWebViewConfiguration = {
+        let configuration = WKWebViewConfiguration()
+        configuration.processPool = ArticleContainerViewController.webProcessPool
+        configuration.setURLSchemeHandler(schemeHandler, forURLScheme: schemeHandler.scheme)
+        return configuration
+    }()
+    
+    lazy var webView: WKWebView = {
+        return WKWebView(frame: view.bounds, configuration: webViewConfiguration)
+    }()
+    
+    
+    // MARK: Lead Image
+    
+    @objc func userDidTapLeadImage() {
+        
+    }
+    
+    func loadLeadImage(with leadImageURL: URL) {
+        leadImageHeightConstraint.constant = leadImageHeight
+        leadImageView.wmf_setImage(with: leadImageURL, detectFaces: true, onGPU: true, failure: { (error) in
+            DDLogError("Error loading lead image: \(error)")
+            self.showWebView()
+        }) {
+            self.layoutLeadImage()
+            self.showWebView()
```

Comment:
Seems like showWebView could be called twice - once when setup() completes and once when lead image loading completes. That might be intentional but just wanted to point out.

Reply:
It was intended (so it's shown when either the image or content are showing) and there's a check at the beginning of the method where it should bail the second time. However, thinking about it now, this was all a workaround for issues with how the old article view worked so it looks like it can be removed for now and we can revisit if the transition breaks again or we want to add the cross fade back in